doc-content/drools-docs/src/main/asciidoc/DecisionEngine/inference-and-truth-maintenance.adoc

```diff
@@ -0,0 +1,271 @@
+[id='inference-and-truth-maintenance_{context}']
+
+= Inference and truth maintenance in the {DECISION_ENGINE}
+
+The basic function of the {DECISION_ENGINE} is to match production facts and data to rules and determine whether and how rules are executed. To ensure that relevant data is applied to the appropriate rules, the {DECISION_ENGINE} makes inferences based on existing knowledge and performs the actions based on the inferred information.
+
+For example, the following DRL rule determines the age requirements for adults, such as in a bus pass policy:
+
+.Rule to define age requirement
+[source]
+----
+rule "Infer Adult"
+when
+  $p : Person(age >= 18)
+then
+  insert(new IsAdult($p))
+end
+----
+
+Based on this rule, the {DECISION_ENGINE} infers whether a person is an adult or a child and performs the specified action (the `then` consequence). Every person who is 18 years old or older has an instance of `IsAdult` inserted for them in the working memory. This inferred relation of age and bus pass can then be invoked in any rule, such as in the following rule segment:
+
+[source]
+----
+$p : Person()
+IsAdult(person == $p)
+----
+
+In many cases, new data in a rule system is the result of other rule executions, and this new data can affect the execution of other rules. If the {DECISION_ENGINE} asserts data as a result of executing a rule, the {DECISION_ENGINE} uses truth maintenance to justify the assertion and enforce truthfulness when applying inferred information to other rules. Truth maintenance also helps to identify inconsistencies and to handle contradictions. For example, if two rules are executed and result in a contradictory action, the {DECISION_ENGINE} chooses the action based on assumptions from previously calculated conclusions.
+
+The {DECISION_ENGINE} inserts facts using either stated or logical insertions:
+
+* *Stated insertions:* Defined with `insert()`. After stated insertions, facts are generally retracted explicitly.
+* *Logical insertions:* Defined with `insertLogical()`. After logical insertions, the facts that were inserted are automatically retracted when the conditions that inserted the facts are no longer true. The facts are retracted when no condition supports the logical insertion.
+
+The following flowcharts illustrate the life cycle of stated and logical insertions:
+
+.Stated insertion
+image::HybridReasoning/Stated_Assertion.png[]
```

Comment:
the term JUSTIFIED should be at least defined somewhere

Reply:
Great point. I also dug this up in an older Drools version, doctored it up, and added it at the end of this section, following the flowcharts. Is this correct and good with you?

>The {DECISION_ENGINE} uses a `HashMap` and a counter to track how many times an object equality is stated (how many different instances are equal). When the {DECISION_ENGINE} logically inserts an object during a rule execution, the {DECISION_ENGINE} _justifies_ the object by executing the rule. For each logical insertion, only one equal object can exist, and each subsequent equal logical insertion increases the justification counter for this logical insertion. A justification is removed when the conditions of the rule become untrue, and the counter is decreased accordingly. When no more justifications exist, the logical object is automatically retracted.

Reply:
I think it is correct, but it's probably too detailed. We can probably be happy with the change to the definition of logical insertion: we can just say there that a fact that has been logically inserted is said to be "justified"

Reply:
Okay, I added the last sentence to this:

>* *Logical insertions:* Defined with `insertLogical()`. After logical insertions, the facts that were inserted are automatically retracted when the conditions that inserted the facts are no longer true. The facts are retracted when no condition supports the logical insertion. A fact that is logically inserted is considered to be _justified_ by the {DECISION_ENGINE}.

And I think we should still keep that detailed bit after the flow charts, don't you? I know it's detailed, but it might be helpful for someone who wants to know more about justification. If they don't, they just won't read it. But if you're not certain about the technical accuracy or feel it's a bad choice to include, I'll kill it.

Let me know.